requests-for-comments/2022-07-14-design-system.md

```diff
@@ -0,0 +1,46 @@
+# Request for comments: PostHogDesign System (Ben White)
+
+## Problem statement
+*Who are we building for, what are their needs (include example cases), why is this important?*
+
+PostHog is a teenage codebase who has gone through lots of infant phases, stumbling blocks and covering itself in [Ants](https://ant.design) for fun. Now that it is growing up, it wants a wardrobe of clothes that match, a bedroom where it can actually find what it is looking for and for people to recognize it's work as distinctly "PostHog". Anyways enough of this analogy...
+
+To keep our product development pace high whilst also keeping our sprawling frontend maintainable we have agreed that it makes sense to solidify some of the great work already done in standardising our code (Lemon Components, utility classes). To do this we need to agree on
+1. What should our Design System cover and what should it not?
+2. How do we ensure Designers and Developers talk in the same language?
+
+
+## Success criteria
+*How do we know if this is successful (i.e. metrics, customer feedback), what's out of scope, whats makes this ambitious?*
+
+Q3 Objective: `50% reduction in custom and inline CSS styles.`
+
+A clear goal is that Designing and Developing the frontend should be faster, perhaps through it being more restrictive. Developers touching custom CSS files should be a rare occurence, both because the components provide all the flexibility required **and** because Designers are also not deviating from the agreed Design System options unless truly necessary.
+
+
+## Context
+*What are our competitors doing, what are the technical constraints, what are customers asking for, what does the data tell us, are there external motivations (e.g. launch week, enterprise contract)?*
+
+> TODO: Link to some good resources on pitfalls of UI library / Design System creation
+
+### LemonUI Components
+
+> TODO: List our current components and any others we agree should be core (✅ - implemented, 🤔 - partially implented, 💡 - not implemented yet)
+
+### Possible rules we could set (remember - this is a proposal to be commented on!)
+
+* Components never layout themselves - no margin or padding
+* Components never expose hacks (e.g. style properties)
+* Anything that isn't a Lemon Component should be styled only with utility classes
+
+### Technical questions
+* Do we continue building this free-form or do we look into using some base Design System library?
+* Do we consider something like Tailwind to aid with tooling, autocomplete, Figma-linking etc?
```

Comment:
+1 to Tailwind

Reply:
I haven't used Tailwind in a professional setting, so these fears might be unwarranted, but...

Can we get by without using Tailwind, but with using our own subset of tailwind-inspired custom classes? I fear that adding all of Tailwind will either add as much CSS as antd, or require a custom CSS tree shaker plugin to get small bundles. The latter will require some non-standard approaches, as we use many JS bundlers (esbuild, webpack), and plan to use even more (vite?). Thus I'm greatly in favour of not adding more complexity to the build system.

Reply:
The important point for me is that we follow Tailwind's API regardless of whether we use Tailwind or not.
As it is so well known (and the docs are great) it would be silly to have utility classes that are custom because it's way too much mental overhead. 

Tailwind itself has lots of great stuff such as auto-completion, breakpoints `sm:foo md:bar` etc. which would be worth getting in but isn't necessarily massively dependent on it.

There are plenty of ways to use Tailwind effectively so it may simply make sense to "try it out" and see what the damage is.